plugin/storage/es/samplingstore/storage.go

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2024 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package samplingstore
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/olivere/elastic"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/cmd/collector/app/sampling/model"
+	"github.com/jaegertracing/jaeger/pkg/es"
+	"github.com/jaegertracing/jaeger/plugin/storage/es/samplingstore/dbmodel"
+)
+
+const (
+	samplingIndex        = "jaeger-sampling-"
+	throughputType       = "throughput-sampling"
+	probabilitiesType    = "probabilities-sampling"
+	indexPrefixSeparator = "-"
+)
+
+type SamplingStore struct {
+	client                         func() es.Client
+	logger                         *zap.Logger
+	samplingIndexPrefix            string
+	indexDateLayout                string
+	maxDocCount                    int
+	samplingIndexRolloverFrequency time.Duration
+	adaptiveSamplingLookback       time.Duration
+}
+
+type SamplingStoreParams struct {
+	Client                 func() es.Client
+	Logger                 *zap.Logger
+	IndexPrefix            string
+	IndexDateLayout        string
+	IndexRolloverFrequency time.Duration
+	Lookback               time.Duration
+	MaxDocCount            int
+}
+
+func NewSamplingStore(p SamplingStoreParams) *SamplingStore {
+	return &SamplingStore{
+		client:                         p.Client,
+		logger:                         p.Logger,
+		samplingIndexPrefix:            p.prefixIndexName(),
+		indexDateLayout:                p.IndexDateLayout,
+		maxDocCount:                    p.MaxDocCount,
+		samplingIndexRolloverFrequency: p.IndexRolloverFrequency,
+		adaptiveSamplingLookback:       p.Lookback,
+	}
+}
+
+func (s *SamplingStore) InsertThroughput(throughput []*model.Throughput) error {
+	ts := time.Now()
+	writeIndexName := indexWithDate(s.samplingIndexPrefix, s.indexDateLayout, ts)
+	s.writeThroughput(writeIndexName, ts, throughput)
+	return nil
+}
+
+func (s *SamplingStore) GetThroughput(start, end time.Time) ([]*model.Throughput, error) {
+	ctx := context.Background()
+	indices := getReadIndices(s.samplingIndexPrefix, s.indexDateLayout, start, end, s.samplingIndexRolloverFrequency)
+	searchResult, err := s.client().Search(indices...).
+		Size(s.maxDocCount).
+		Query(buildTSQuery(start, end)).
+		IgnoreUnavailable(true).
+		Do(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to search for throughputs: %w", err)
+	}
+	var retSamples []dbmodel.Throughput
```

Comment:
you can pre-allocate based on `len(searchResult.Hits.Hits)` and assign by position instead of using append

Reply:
I think append will be better because both are arrays; if we assign by position, we have to create a 2d array, and then convert this 2d array of dbmodel throughput into a 1d array of model throughput.

Reply:
not sure what you mean. I am talking about this pattern:
```
input := ...
output = make([]Output, len(input))
for i, v := range input {
  json.Unmarshal(v, &output[i])
}
```
In this style you pre-allocate the output array of exactly the size you need, which can be more efficient than doing append and potentially having to extend the slice. You can also unmarshal directly into array element (if the slice is defined over values, not pointers) and avoid extra copying.

Btw, in `retSamples = append(retSamples, tToD.Throughput...)` the `...` doesn't make sense since you're appending a single element, not another slice.

Reply:
> Btw, in retSamples = append(retSamples, tToD.Throughput...), the ... doesn't make sense since you're appending a single element, not another slice.

tTod.Throughput is also a slice
```go
type TimeThroughput struct 
    Timestamp  time.Time    `json:"timestamp"`
    Throughput []Throughput `json:"throughputs"`
}
```
I agree that it may create extra copies, which may cause vulnerabilities. working on it

Reply:
I think my current approach is wrong. Currently I am inserting throughputs like one timestamp-> array of throughputs,
```go
type TimeThroughput struct {
	Timestamp  time.Time    `json:"timestamp"`
	Throughput []Throughput `json:"throughputs"`
}
```
maybe we want something like this :
```go
type TimeThroughput struct {
	Timestamp  time.Time    `json:"timestamp"`
	Throughput Throughput   `json:"throughput"`
}
```